inmoose/cohort_qc/cohort_metric.py

```diff
@@ -0,0 +1,864 @@
+            )
+        else:
+            score_before = None
+
+        return score_before, score_after
+
+    def compute_entropy(
+        self, data_expression_df: pd.DataFrame, n_neighbors: int
+    ) -> float:
+        """
+        Compute the entropy of batch mixing to evaluate the effectiveness of batch correction.
+
+        Parameters
+        ----------
+        data_expression_df : pandas.DataFrame
+            Gene expression data frame to evaluate.
+        n_neighbors : int
+            Number of nearest neighbors to consider in the entropy calculation.
+
+        Returns
+        -------
+        float
+            Mean entropy value indicating the level of batch mixing.
+        """
+
+        nbrs = NearestNeighbors(n_neighbors=n_neighbors, metric="euclidean").fit(
+            data_expression_df.T
+        )
+        _, indices = nbrs.kneighbors(data_expression_df.T)
+        batch_labels = self.clinical_df[self.batch_column].values
+
+        entropies = []
+        for idx in range(indices.shape[0]):
+            neighbors_labels = batch_labels[indices[idx]]
+            label_counts = np.unique(neighbors_labels, return_counts=True)[1]
+            probabilities = label_counts / len(neighbors_labels)
+            entropy = -np.sum(probabilities * np.log2(probabilities))
+            entropies.append(entropy)
+
+        return np.mean(entropies)
+
+    def entropy_batch_mixing(self, n_neighbors: int = 5) -> tuple:
+        """
+        Compute the entropy of batch mixing before and after batch correction.
+
+        Parameters
+        ----------
+        n_neighbors : int, optional
+            Number of nearest neighbors to consider in the entropy calculation, by default 5.
+
+        Returns
+        -------
+        tuple
+            Entropy values before and after correction.
+        """
+        # Compute entropy of batch mixing after batch correction
+        entropy_after = self.compute_entropy(self.data_expression_df, n_neighbors)
+
+        if self.data_expression_df_before is not None:
+            # Compute entropy of batch mixing before batch correction
+            entropy_before = self.compute_entropy(
+                self.data_expression_df_before, n_neighbors
+            )
+        else:
+            entropy_before = None
+
+        return entropy_before, entropy_after
+
+    def summarize_and_compare_mixed_datasets(self) -> dict:
+        """
+        Summarize and compare mixed datasets to non-mixed datasets to evaluate their representativeness.
+
+        Returns
+        -------
+        dict
+            Summary with details about each mixed dataset and all mixed datasets in general.
+        """
+
+        summary = {}
+
+        # Identify mixed and non-mixed datasets
+        non_mixed_datasets = self.clinical_df[
+            ~self.clinical_df[self.batch_column].isin(self.mixed_datasets)
+        ][self.batch_column].unique()
```

Comment:
What about a setdiff between self.clinical_df[self.batch_column].values and self.mixed_datasets?